60 handle errors from non http url sources

EddyCMWF opened the pull request:
ftp is not compatible with multiurl, so if ftp we use the wget downloader.
Hopefully the approach is generic to facilitate other downloaders in the future.

malmans2 commented on tests/test_10_tools.py:
```diff
@@ -0,0 +1,38 @@
+@pytest.fixture()
+def ftp(tmp_path):
+    ftp_path = tmp_path / "ftp"
+    (ftp_path / "test").mkdir(parents=True)
```
@EddyCMWF since you are revising `try_download`, I think there's another bug.

If I don't place the file in a folder, `try_download` fails here: https://github.com/ecmwf-projects/cads-adaptors/blob/963cb049b810ae8eff85c3387d658486fec1eee5/cads_adaptors/tools/url_tools.py#L36-L37

EddyCMWF commented:
Hi Mattia,
I've fixed and released multiurl which addresses the issue upstream of cads-adaptors.
I've also added an ftp download test.
So good for a re-review.
Eddy

malmans2 commented:
I'll take a look tomorrow (bank holiday in Italy today).
I you didn't notice it, the other day I opened a PR to this branch to test both ftp ad http: https://github.com/ecmwf-projects/cads-adaptors/pull/65
It also fixes this https://github.com/ecmwf-projects/cads-adaptors/pull/61#discussion_r1375806860

malmans2 commented on .gitignore:
```diff
@@ -471,3 +471,5 @@ $RECYCLE.BIN/
 # End of https://www.toptal.com/developers/gitignore/api/python,jupyternotebooks,vim,visualstudiocode,pycharm,emacs,linux,macos,windows
 test.ipynb
+
+testfile.txt
```
IMO test should not produce any permanent file.
So they should either be locate in a temporary directory (i.e., using the `tmp_path` fixture) or removed after the tests are exectuted (pytest has teardown capabilities).

Also, I would place additions to the default gitignore in the header of this file, where there are others additions to the default of `toptal`. Otherwise, these additions are very hard to find for other developers.

EddyCMWF replied:
done

malmans2 commented on cads_adaptors/tools/url_tools.py:
```diff
@@ -8,6 +8,7 @@
 import jinja2
 import multiurl
 import requests
+import yaml
```
pyyaml is not python standard library. Looks like another dependency is installing it, but that's not reliable.
If you need to use pyyaml, you should add it to the`environment.yml` and to the list of mandatory dependencies in `pyproject.toml`. It's only used to format an error message though. Is it worth it?

EddyCMWF replied:
nxt commit

malmans2 commented on cads_adaptors/tools/url_tools.py:
```diff
@@ -35,21 +36,20 @@ def try_download(urls: List[str]) -> List[str]:
         try:
             multiurl.download(url, path)
+        except Exception as exc:
+            excs.append({url: exc})
```
This list was populated to show all the errors in the RuntimeError message, but that feature is removed in this PR.
I think you should either restore that feature, or get rid of the `excs` list as well.

EddyCMWF replied:
This was intentional to try and make the error messages received by users more undertandable, but you're right, I no longer need the excs list as we now dump those to internal logs straight away

malmans2 commented on cads_adaptors/tools/url_tools.py:
```diff
@@ -35,21 +36,20 @@ def try_download(urls: List[str]) -> List[str]:
-            f"Request empty. At least one of the following {urls} "
-            "must be a valid url from which to download the data "
-            f"download errors: {[str(exc) for exc in excs]}"
+            f"Request empty. At least one of the following:\n{yaml.safe_dump(urls, indent=2)} "
```
Two ideas:
1. Avoid using an external package: `"  "+"\n  ".join(urls)`
2. If you want to restore the display of all exceptions, python3.11 introduces ExceptionGroup (we are already testing py3.11, we will eventually move from 3.10 to 3.11). Here is an example from cacholote: https://github.com/ecmwf-projects/cacholote/blob/bf87b56a7f160de61ab644932b50ad66515accc3/cacholote/encode.py#L164-L166

EddyCMWF replied:
I think error messaging layout can be addressed in a separate ticket, let's just get this merged so it's working

malmans2 commented on ci/environment-ci.yml:
```diff
@@ -17,5 +17,6 @@ dependencies:
 - types-python-dateutil
 - types-pyyaml
 - types-requests
+- pytest-localftpserver
```
I did consider using it, but I'm a bit worried that this is a personal project rather than a reliable library.
It only has a dozen of start and a few forks. Furthermore, it does not support macOS.
Not sure if you use macOS (most of us do at B-Open), but I get a warning and it looks like the issue is not going to be fixed: `https://github.com/oz123/pytest-localftpserver/issues/151#issuecomment-922530735`

EddyCMWF replied:
I'm using macOS and it's running fine for me
If it becomes troublesome we can swap it out later

malmans2 commented on tests/test_20_url_tools.py:
```diff
@@ -0,0 +1,45 @@
+from cads_adaptors.tools import url_tools
+
+
+@pytest.mark.parametrize(
+    "urls,expected_nfiles",
+    (
+        (
+            [
+                "https://get.ecmwf.int/repository/test-data/earthkit-data/test-data/test_single.nc"
+            ],
+            1,
+        ),
+        (
+            [
+                "https://get.ecmwf.int/repository/test-data/earthkit-data/test-data/test_single.nc",
+                "https://get.ecmwf.int/repository/test-data/earthkit-data/test-data/test_single.grib",
+            ],
+            2,
+        ),
+    ),
+)
```
A couple of comments about this parametrization:
1. The internal logic of url_tools does not make any difference between len(1) or len(N>1) urls.
2. Why testing just the length of the output. You know exactly the output expected right?

EddyCMWF replied:
It was just to ensure that the handling of single files and multiple files was preserved.
Yes more can be added, but this is better than nothing

malmans2 replied:
In case it wasn't clear, I wasn't suggesting to add more tests. Just this:
```python
@pytest.mark.parametrize(
    "urls,expected_files",
    (
        (
            [
                "https://get.ecmwf.int/repository/test-data/earthkit-data/test-data/test_single.nc"
            ],
            ["repository/test-data/earthkit-data/test-data/test_single.nc"],
        ),
        (
            [
                "https://get.ecmwf.int/repository/test-data/earthkit-data/test-data/test_single.nc",
                "https://get.ecmwf.int/repository/test-data/earthkit-data/test-data/test_single.grib",
            ],
            [
                "repository/test-data/earthkit-data/test-data/test_single.nc",
                "repository/test-data/earthkit-data/test-data/test_single.grib",
            ],
        ),
    ),
)
def test_downloaders(tmp_path, monkeypatch, urls, expected_files):
    monkeypatch.chdir(tmp_path)  # try_download generates files in the working dir
    assert url_tools.try_download(urls) == expected_files
```

malmans2 commented on tests/test_20_url_tools.py:
```diff
@@ -0,0 +1,45 @@
+    with open(local_test_file, "w") as f:
+        f.write("This is a test file")
+
+    ftp_url = ftpserver.put_files(local_test_file, style="url", anon=True)
+    local_test_download = url_tools.try_download(ftp_url)[0]
+    with open(local_test_file) as original, open(local_test_download) as downloaded:
+        assert original.read() == downloaded.read()
+
+    ftp_url = ftpserver.put_files(local_test_file, style="url", anon=False)
+    local_test_download = url_tools.try_download(ftp_url)[0]
+    with open(local_test_file) as original, open(local_test_download) as downloaded:
+        assert original.read() == downloaded.read()
```
Is this code just repeated twice? Why? If it's to test that it's able to overwrite existing files I would make it more clear.

malmans2 replied:
Ohhh I see, anon changed. I think you should use `@pytest.mark.parametrize` then.

EddyCMWF replied:
in nxt commit

malmans2 reviewed (approved):
Hi @EddyCMWF,

I have a few comments. I'm approving so you can merge when you are happy with it.

About the tests: As I mentioned in the comments, I'm a bit worried that the library used in this PR is not stable/reliable.
Also, IMO it's better if we don't rely on external files. That way we don't need internet access to run the tests, we don't download external files every time tests are executed, and we are not affected by the external data (e.g., urls could change, downtimes, ...)
Given that the ftp fixture is just about 10 lines of code, I've aligned another branch to this PR in case you change your mind and you want to use my version of the tests: https://github.com/ecmwf-projects/cads-adaptors/pull/66

malmans2 commented on tests/test_20_url_tools.py:
```diff
@@ -0,0 +1,45 @@
+
+
+def test_ftp_download(tmp_path, ftpserver):
+    local_test_file = os.path.join(tmp_path, "testfile.txt")
+    with open(local_test_file, "w") as f:
+        f.write("This is a test file")
```
`pathlib` replaces `os`, and `tmp_path ` is a `pathlib.Path object`.
```suggestion
    local_test_file = tmp_path / "testfile.txt"
    local_test_file.write_text("This is a test file")
```

Also below you can use methods such as `.read_text`

EddyCMWF replied:
okay, I'm fine with the ol' skool method here

malmans2 commented on environment.yml:
```diff
@@ -17,4 +17,4 @@ dependencies:
 - pip:
   - rooki
-  - multiurl
+  - multiurl>=0.2.3.2
```
multiurl is on conda-forge, and `0.2.3.2` is already available so I would install using conda rather than pip
